docs/design/draft/initial-design.md

```diff
@@ -0,0 +1,379 @@
+explicitly to PVC configuration can probably be kept as explicit CRD
+fields, though, since any data mover implementation will need to deal
+with them.
+
+per-PVC spec parameters
+- `originalPVC` (namespaced name)
+- `targetPVC` (namespaced name)
+  Optional: if not specified, use `originalPVC` values
+- `targetStorageClassName`
+  Optional: if not specified, use storageClassName from backup
+- `targetAccessModes`
+  Optional: if not specified, use accessModes from backup
+- `targetStorage`
+  Optional: Used for setting `Requests.Storage` for the new PVC. If
+  omitted, the value from `Capacity.Storage` from the PVC backed up is used
+
+A PVC list will be generated in the `Status` field with omitted values
+from the Spec field filled in with the from-backup defaults.
+
+Here is an example of a `DataMoverRestore` resource.
+
+```
+apiVersion: oadp.openshift.io/v1alpha1
+kind: DataMoverRestore
+metadata:
+  name: example-restore
+  namespace: oadp-example
+spec:
+  dataMoverBackup: example-backup
+# if UIDs are used:
+# dataMoverBackup: fbd0eeeb-6ea7-11ea-912c-0274cdd9b69e
+  storageLocationRef:
+    name: storage-example
+    namespace: oadp-example
+  pvcs:
+  - originalPVC:
+      name: mysql
+      namespace: mysql-persistent
+    targetPVC:
+      name: mysql-moved
+      namespace: mysql-persistent
+    targetStorageClassName: csi-ceph
+    targetAccessModes:
+    - ReadWriteMany
+  - originalPVC:
+      name: mysql2
+      namespace: mysql-persistent
+    targetStorage: 20Gi
+status:
+  pvcs:
+  - originalPVC:
+      name: mysql
+      namespace: mysql-persistent
+    targetPVC:
+      name: mysql-moved
+      namespace: mysql-persistent
+    targetStorageClassName: csi-ceph
+    targetAccessModes:
+    - ReadWriteMany
+    targetStorage: 10Gi
+    # This field may need to belong to a map rather than as an
+    # explicit field since it's plugin-specific
+    resticSnapshotId: b2493502
+  - originalPVC:
+      name: mysql2
+      namespace: mysql-persistent
+    targetPVC:
+      name: mysql2
+      namespace: mysql-persistent
+    targetStorageClassName: csi-ebs
+    targetAccessModes:
+    - ReadWriteOnce
+    targetStorage: 20Gi
+    resticSnapshotId: c805056d
+
+```
+
+## Data Mover Controller
+
+The Data Mover Controller is a long-running pod (implemented as a
+`DeploymentConfig`) which watches `DataMoverBackup` and
+`DataMoverRestore` resources.
+
+TBD: The Backup and Restore CRs will have a status/phase field with a
+to-be-determined workflow. Possible states to deal with include:
+
+- New (might not be needed since once Status is no longer empty, we've
+  moved beyond this status
+- Processing (sub-states? progress reporting?)
+- JobRunning
+- DeletingPVCs (Backup only)
+- Completed
+- Failed (sub-states? FailedProcessing, FailedBackupJob, etc.)
+
+### Backups
+
+For each new `DataMoverBackup`:
+- For each PVC to be backed up, find a VolumeSnapshot
+  - If no snapshot, do we fail? skip? create one?
```

Comment:
Creating one would be cool... but optionally if we find a snapshot and it's old how would we know if we wanted to take a new one?

Reply:
Yeah, that's why my original version of this included creating across the board. But since users may already have their snapshot-generation infrastructure in place, it probably shouldn't be the default. Maybe for v1 of this we don't create snapshots and assume the user has created them. Skip vs. fail remains a question. If we decide to add snapshot creation, we'd need some more CRD params:
1) create/ignore/fail choice for snapshot not found
2) max snapshot age to consider (older should be equivalent to not found)
3) possibly other params needed for snapshot creation